core/src/main/java/io/questdb/griffin/engine/analytic/CachedAnalyticRecordCursorFactory.java

```diff
@@ -104,92 +106,156 @@ protected void _close() {
+            final Record chainRightRecord = cursor.getRecordB();
+            if (orderedGroupCount > 0) {
+                while (base.hasNext()) {
+                    offset = recordChain.put(record, offset);
                     recordChain.recordAt(recordChainRecord, offset);
-                    for (int j = 0; j < functionCount; j++) {
-                        functions.getQuick(j).pass1(recordChainRecord, offset, recordChain);
+                    for (int i = 0; i < orderedGroupCount; i++) {
+                        orderedSources.getQuick(i).put(recordChainRecord, recordChain, chainRightRecord, comparators.getQuick(i));
                     }
                 }
+            } else {
+                while (base.hasNext()) {
+                    offset = recordChain.put(record, offset);
+                }
             }
-        }
 
-        // run pass1 for all unordered functions
-        if (unorderedFunctions != null) {
-            for (int j = 0, n = unorderedFunctions.size(); j < n; j++) {
-                final AnalyticFunction f = unorderedFunctions.getQuick(j);
-                recordChain.toTop();
-                while (recordChain.hasNext()) {
-                    f.pass1(recordChainRecord, recordChainRecord.getRowId(), recordChain);
+            if (orderedGroupCount > 0) {
+                for (int i = 0; i < orderedGroupCount; i++) {
+                    final LongTreeChain tree = orderedSources.getQuick(i);
+                    final ObjList<AnalyticFunction> functions = orderedFunctions.getQuick(i);
+                    // step #2: populate all analytic functions with records in order of respective tree
+                    final LongTreeChain.TreeCursor cursor = tree.getCursor();
+                    final int functionCount = functions.size();
+                    while (cursor.hasNext()) {
+                        offset = cursor.next();
+                        recordChain.recordAt(recordChainRecord, offset);
+                        for (int j = 0; j < functionCount; j++) {
+                            functions.getQuick(j).pass1(recordChainRecord, offset, recordChain);
+                        }
+                    }
                 }
             }
+
+            // run pass1 for all unordered functions
+            if (unorderedFunctions != null) {
+                for (int j = 0, n = unorderedFunctions.size(); j < n; j++) {
+                    final AnalyticFunction f = unorderedFunctions.getQuick(j);
+                    recordChain.toTop();
+                    while (recordChain.hasNext()) {
+                        f.pass1(recordChainRecord, recordChainRecord.getRowId(), recordChain);
+                    }
+                }
+            }
+
+            recordChain.toTop();
         }
 
-        recordChain.toTop();
-        return recordChain;
-    }
+        @Override
+        public void close() {
+            base.close();
+            recordChain.close();
```

Comment:
We're sinking down in this murky semantics that close() doesn't actually make object unusable.

RecordCursor is not "mutable", so it doesn't have clear "clear()" method. Instead it is sort of mutable with `toTop` call.

The semantics of what we do here is now a bit of spaghetti. The intent behind closing cursor and opening new one is that next open will re-use memory allocated by previous runs and thus avoid free/alloc calls.

`clear()` method will be appropriate here. But for that to make sense `chain` should not be owned by the cursor, but rather factory. This is how all the reusable state is held, or should be held

Reply:
Going to address this later once I fix variations of the problematic query as @ideoma pointed out. These are:
```sql
select row_number() over (partition by symbol order by symbol), * from trades;
select row_number() over (partition by symbol order by symbol) from trades;
```

Reply:
Tried to address the suggestion in f6fdca3e9ea34a02ae499cc52612ef788bd62f3d